src/coreclr/vm/jithelpers.cpp

```diff
@@ -2438,109 +2444,180 @@ void JIT_Patchpoint(int* counter, int ilOffset)
-        // for OSR methods, so we stick with synchronous creation.
-        LOG((LF_TIEREDCOMPILATION, LL_INFO10, "Jit_Patchpoint: patchpoint [%d] (0x%p) TRIGGER at count %d\n", ppId, ip, hitCount));
+    if (osrMethodCode == (PCODE)NULL)
+    {
+        MAKE_CURRENT_THREAD_AVAILABLE();
+
+    #ifdef _DEBUG
+        Thread::ObjectRefFlush(CURRENT_THREAD);
+    #endif
 
-        // Invoke the helper to build the OSR method
-        osrMethodCode = HCCALL3(JIT_Patchpoint_Framed, pMD, codeInfo, ilOffset);
+        FrameWithCookie<DynamicHelperFrame> frame(pTransitionBlock, 0);
+        DynamicHelperFrame * pFrame = &frame;
+
+        pFrame->Push(CURRENT_THREAD);
+
+        INSTALL_MANAGED_EXCEPTION_DISPATCHER;
+        INSTALL_UNWIND_AND_CONTINUE_HANDLER;
+
+        GCX_PREEMP();
+
+        osrMethodCode = ppInfo->m_osrMethodCode;
+        if ((osrMethodCode == (PCODE)NULL) && forceTransition)
+        {
+            // Another thread has already triggered the OSR method, wait until it is available, or the other thread abandons trying to JIT the OSR method
+            DWORD backoffs = 0;
+            while (osrMethodCode = ppInfo->m_osrMethodCode, osrMethodCode == (PCODE)NULL)
+            {
+                LONG oldFlags = ppInfo->m_flags;
+                if ((oldFlags & PerPatchpointInfo::patchpoint_triggered) == PerPatchpointInfo::patchpoint_triggered)
+                {
+                    LOG((LF_TIEREDCOMPILATION, LL_INFO1000, "Jit_PartialCompilationPatchpoint: AWAITING OSR method for patchpoint [%d] (0x%p)\n", ppId, ip));
+                    __SwitchToThread(0, backoffs++);
+                    continue;
+                }
+
+                LONG newFlags = ppInfo->m_flags | PerPatchpointInfo::patchpoint_triggered;
+                BOOL triggerTransition = InterlockedCompareExchange(&ppInfo->m_flags, newFlags, oldFlags) == oldFlags;
+
+                if (!triggerTransition)
+                {
+                    LOG((LF_TIEREDCOMPILATION, LL_INFO1000, "Jit_PartialCompilationPatchpoint: (lost race) AWAITING OSR method for patchpoint [%d] (0x%p)\n", ppId, ip));
+                    __SwitchToThread(0, backoffs++);
+                    continue;
+                }
+            }
+        }
 
-        // If that failed, mark the patchpoint as invalid.
         if (osrMethodCode == (PCODE)NULL)
         {
-            // Unexpected, but not fatal
-            STRESS_LOG3(LF_TIEREDCOMPILATION, LL_WARNING, "Jit_Patchpoint: patchpoint (0x%p) OSR method creation failed,"
-                " marking patchpoint invalid for Method=0x%pM il offset %d\n", ip, pMD, ilOffset);
+            // Time to create the OSR method.
+            //
+            // We currently do this synchronously. We could instead queue
+            // up a request on some worker thread, like we do for
+            // rejitting, and return control to the Tier0 method. It may
+            // eventually return here, if the patchpoint is hit often
+            // enough.
```

Comment:
This part of the comment does not make sense for partial compilation, the current thread is blocked until the code becomes available.

Reply:
Resolved as this comment is now only in the non-partial compilation policy handler.